sql/src/main/java/io/crate/operation/merge/SortingBucketMerger.java

```diff
@@ -104,50 +91,44 @@ public SortingBucketMerger(RowDownstream rowDownstream,
      * output:
      *  [ A, A, B, B, B, B, C, C, C, D, D ]
      *
-     * see private emitBuckets(...) for more details on how this works
      */
     @Override
     public void nextPage(BucketPage page, final PageConsumeListener listener) {
         assert page.buckets().size() == numBuckets :
                 "number of buckets received in merge call must match the number given in the constructor";
-        final AtomicBoolean listenerNotified = new AtomicBoolean(false);
 
         FutureCallback<List<Bucket>> finalCallback = new FutureCallback<List<Bucket>>() {
             @Override
             public void onSuccess(List<Bucket> buckets) {
-                try {
-                    ArrayList<Iterator<Row>> bucketIts = new ArrayList<>(numBuckets);
-                    for (int i = 0; i < buckets.size(); i++) {
-                        Iterator<Row> remainingBucketIt = remainingBucketIts[i];
-                        if (remainingBucketIt == null) {
-                            bucketIts.add(buckets.get(i).iterator());
-                        } else {
-                            bucketIts.add(Iterators.concat(remainingBucketIt, buckets.get(i).iterator()));
-                        }
+                List<Iterator<Row>> bucketIts = getBucketIterators(buckets);
+                RowMergingIterator mergingIterator = getMergingIterator(bucketIts);
+
+                while (mergingIterator.hasNext()) {
+                    if (mergingIterator.leastExhausted()) {
+                        prevRowMergingIterator = mergingIterator;
+                        break;
+                    }
+
+                    if (finished.get()) {
```

Comment:
shouldn't this be the first check while iterating? otherwise it could be never reached or?